### ISSUE TITLE
[HttpClient] Add a method to validate the status code of the response

### ISSUE THREAD

**Issue by mnapoli** [symfony/symfony#34046]
**Description**  

Checking all error cases of a response is repetitive work. I understand why the HTTP client does not throw exceptions on HTTP errors.

However, a helper method could make things simpler for most cases: a new `$response->assertIsSuccessful()` method could throw detailed errors if the HTTP status code is >=300.

**Example**  

```php
// Before
$response = $http->request('PATCH', '<url>', ...);
if ($response->getStatusCode() === 404) {
    throw new NotFound('Resource not found');
}
if ($response->getStatusCode() === 400) {
    throw new RuntimeException('We need to fix our code : ' /* add info about the response here */);
}
if ($response->getStatusCode() >= 500) {
    throw new RuntimeException('API error, we need to retry later');
}
// and I'm missing some cases here

// After
$response = $http->request('PATCH', '<url>', ...);
// Throws if response is not 200
$response->assertIsSuccessful();
```

From [the documentation](https://symfony.com/doc/current/components/http_client.html#handling-exceptions) it seems the only way to achieve a similar behavior is to try to get the content of the response. But it would make more sense to have a method dedicated to that.

WDYT?

**Comment by nicolas-grekas** [comment]
`$response->getHeaders(true);` does what you want (`true` is the default already but in your case you might want to make it explicit).

**Comment by nicolas-grekas** [comment]
Actually, you don't have to do anything explicit: the exception will be thrown also when doing only
`$http->request('PATCH', '<url>', ...);`

I'm reluctant to adding more methods when they "just" alias something the current API already provides... that makes e.g decorating more difficult, one more method to train ppl for, etc.

**Comment by mnapoli** [comment]
> Actually, you don't have to do anything explicit: the exception will be thrown also when doing only
$http->request('PATCH', '<url>', ...);

That's interesting! Do you have more information about that? I can't find it in the docs. You mean that the response will automatically be validated?

**Comment by nicolas-grekas** [comment]
> You mean that the response will automatically be validated?

Yes! And you're right this would deserve a section in the doc! Please open a doc issue.
